rebasebot/bot.py

```diff
@@ -137,32 +138,65 @@ def _add_to_rebase(commit_message, source_repo, tag_policy):
     logging.info("Performing rebase")
 
     merge_base = gitwd.git.merge_base(f"source/{source.branch}", f"dest/{dest.branch}")
     logging.info("Rebasing from merge base: %s", merge_base)
 
     # Find the list of commits between the merge base and the destination head
     # This should be the list of commits we are carrying on top of the UPSTREAM
-    commits = gitwd.git.log("--reverse", "--pretty=format:%H - %s", "--no-merges",
+    commits = gitwd.git.log("--reverse", "--pretty=format:%H - %s - %aE", "--no-merges",
                             "--ancestry-path", f"{merge_base}..dest/{dest.branch}")
     logging.info("Picking commits: \n%s", commits)
 
+    commits_to_squash = defaultdict(list)
+
     for commit in commits.splitlines():
         # Commit contains the message for logging purposes,
         # trim on the first space to get just the commit sha
-        sha, commit_message = commit.split(" - ")
+        sha, commit_message, committer_email = commit.split(" - ")
 
         if not _add_to_rebase(commit_message, source_repo, tag_policy):
             continue
 
         try:
-            gitwd.git.cherry_pick(f"{sha}", "-Xtheirs")
+            if allow_bot_squash:
+                # There is sometimes a prefix with number and a following + sign
+                # We have to get rid of that part to make sure to get
+                # only the email of the bot
+                is_plus_in_email = committer_email.find("+")
+                email = ""
+                if is_plus_in_email != -1:
+                    _, email = committer_email.split("+")
+                else:
+                    email = committer_email
+                if email in bot_emails:
+                    commits_to_squash[email].append(sha)
+                else:
+                    logging.info(f"cherry-picking {sha}")
+                    gitwd.git.cherry_pick(f"{sha}", "-Xtheirs")
+            else:
+                gitwd.git.cherry_pick(f"{sha}", "-Xtheirs")
         except git.GitCommandError as ex:
             if not _resolve_rebase_conflicts(gitwd):
                 raise RepoException(f"Git rebase failed: {ex}") from ex
 
+    # Here we cherry-pick the bot's commits and then squash them together
+    # We also want the newest bot commit message to represent the squashed commits
+    if allow_bot_squash:
+        for _, value in commits_to_squash.items():
+            try:
+                for sha in value:
+                    gitwd.git.cherry_pick(f"{sha}", "-Xtheirs")
+            except git.GitCommandError as ex:
+                    if not _resolve_rebase_conflicts(gitwd):
+                        raise RepoException(f"Git rebase failed: {ex}") from ex
+            gitwd.git.reset("--soft", f"HEAD~{len(value)}")
+            newest_bot_commit_message = gitwd.git.log("-n 1", f"{value[len(value)-1]}","--pretty=format:%s")
+            print(len(value), newest_bot_commit_message)
+            gitwd.git.commit("-m", newest_bot_commit_message)
```

Comment:
Do we have the email address for the bot available here? You can override the author and set it to the bot email with the `--author` flag, might be nice to allow future rebases to work out that this is a bot commit and can be squashed again

Reply:
Why do we need the email here?
What we do in the code you highlighted is:
1. git reset --soft HEAD~<number_of_bots_commit_per_email>
2. we get the newest message (`-n 1`) by using `value[len(value)-1]` - `value` is a list of all sha's from the bot's commits
3. finally we commit then changed squashed commits

Reply:
When you execute the commit, unless you specify a specific author, the commit will be authored by the rebase bot.
So, when you run the first rebase, we have the ART bot and rebase bot making commits, we then squash all ART bot commits into one, and all rebase bot commits into one. If we don't use the original bot email to author the commits, then when we rebase for a second time, all of the commits from prior rebases that were squashed, are now authored by the rebase bot, so the ART bot content will now get squashed with the rebase bot content, which then defeats the feature of squashing by bot

Reply:
Understood. Makes sense, totally forgot about this. Changes on the way!